src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
+				       sizeof(buf), HID_FEATURE_REPORT,
+				       HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to retrieve macro for profile %d for button %d: %d\n",
+			  profile, button, rc);
+		return NULL;
+	}
+
+	/* Wait for the device to be ready */
+	msleep(100);
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GET_SET_MACRO,
+				       (uint8_t*)report, sizeof(*report),
+				       HID_FEATURE_REPORT, HID_REQ_GET_REPORT);
+	if (rc < (signed)sizeof(*report)) {
+		log_error(device->ratbag,
+			  "Failed to retrieve macro for profile %d for button %d: %d\n",
+			  profile, button, rc);
+		return NULL;
+	}
+
+	checksum = gskill_calculate_checksum((uint8_t*)report, sizeof(*report));
+	if (checksum != report->checksum) {
+		log_error(device->ratbag,
+			  "Invalid checksum on macro for profile %d button %d\n",
+			  profile, button);
+		return NULL;
+	}
+
+	return report;
+}
+
+static int
+gskill_write_button_macro(struct ratbag_device *device,
+			  struct gskill_macro_report *report)
+{
+	unsigned int profile = report->macro_num / 10;
+	unsigned int button = report->macro_num % 10;
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x0b,
+		report->macro_num, 0x1 };
+	int rc;
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
+				       sizeof(buf), HID_FEATURE_REPORT,
+				       HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to set macro to write for profile %d button %d: %d\n",
+			  profile, button, rc);
+		return rc;
+	}
+
+	/* Wait for the device to be ready */
+	msleep(200);
+
+	memset(&report->header, 0, sizeof(report->header));
+	report->header.write.report_id = 0x4;
+	report->checksum = gskill_calculate_checksum((uint8_t*)report,
+						     sizeof(*report));
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GET_SET_MACRO,
+				       (uint8_t*)report, sizeof(*report),
+				       HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to write macro for profile %d button %d to mouse: %d\n",
+			  profile, button, rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+static int
+gskill_probe(struct ratbag_device *device)
+{
+	struct gskill_data *drv_data = NULL;
+	struct ratbag_profile *profile;
+	unsigned int active_idx;
+	int ret;
+
+	ret = ratbag_open_hidraw(device);
+	if (ret)
+		return ret;
+
+	drv_data = zalloc(sizeof(*drv_data));
+	ratbag_set_drv_data(device, drv_data);
+
+	ret = gskill_get_profile_count(device);
+	if (ret < 0)
+		goto err;
+
+	/*
+	 * TODO: Add proper support for enabling/disabling profiles in
+	 * libratbag. For now we workaround this by just setting the profile
+	 * count to 5
+	 */
+	if (ret < GSKILL_PROFILE_MAX) {
+		log_info(device->ratbag,
+			 "We don't support dynamically enabling/disabling profiles yet, sorry! Setting profile count of mouse to 5\n");
+		ret = gskill_set_profile_count(device, GSKILL_PROFILE_MAX);
```

Comment:
If we make sure that the profiles are valid (which should just entail giving them a name, something I'll add on v2) then no, other then setting the max profile count to 5. While this isn't a very nice solution, the mouse throws a fit (e.g. sends us part of it's ram contents) if we try to get profile `n` where `n >= profile_count`. The other solution would be to just set the supported profile count to whatever the max number of profiles is atm, but then we'd give the user no way to add another profile without going into windows.
I'm not too worried about this though, since we won't need this once we have support for enabling/disabling profiles.